### PR TITLE
Allow declaring unintialized ROS parameters

### PR DESCRIPTION
**Pull request by domire8**
Ok wow that was painful. So far, it was not possible to create ROS parameters from an empty state rep parameter. One idea was to do that: 
```
if sr_parameter.is_empty():
    self.declare_parameter(ros_param.name, None,
                           descriptor=ParameterDescriptor(description=description,
                                                          type=Parameter.Type.DOUBLE_ARRAY))
```
This however, does not actually create an uninitialized ROS parameter of type double array, but a ROS parameter of type `NOT_SET`. Which means that next time we want to set that parameter in the on_set_parameter_callback, it will be incompatible. So the solution is that:
```
if sr_parameter.is_empty():
    self.declare_parameter(ros_param.name, None,
                           descriptor=ParameterDescriptor(description=description,
                                                          type=Parameter.Type.NOT_SET.value,
                                                          dynamic_typing=True))
```
We declare the parameter with type NOT_SET and allow dynamic_typing. Now you might say, no wait we dont want dynamic typing. Well it's really only dynamically typed until its set for the first time because every following time that we want to modify the parameter, we are checking that the types correspond, so we are good. This behaviour should be illustrated in the test. We can discuss if thats not what we want.

### CODE REVIEW

File: source/modulo_components/test/python/test_component.py
Comment (eeberhard) [inline source/modulo_components/test/python/test_component.py:36]: maybe you can add some simple comments here to make it clear what the test is expecting. In this case the parameter validation fails and the type (and value) are not changed, but it's not particularly obvious

**Review by eeberhard — COMMENTED** [review]
That's a smart solution! Thanks for taking the time to work it out. 

The consequence is that "default empty" parameters cannot be validated automatically the first time they are set, which is acceptable but will need to be made clear in the documentation later. Any validation would have to occur "in place" where you get and use the parameter value. 

One question before I approve - what happens if we try to get the value of a parameter that is NOT_SET? Internally I suppose it's just the sr.Parameter object or attribute, which will be "empty". But if I were to read that parameter on the ROS interface, what would that look like?

**Comment by domire8** [comment]
> One question before I approve - what happens if we try to get the value of a parameter that is NOT_SET? Internally I suppose it's just the sr.Parameter object or attribute, which will be "empty". But if I were to read that parameter on the ROS interface, what would that look like?

Internally, you get the sr.Parameter which is empty yes but you can still get the value, will be the default empty value of that specific parameter type.
If you do `ros2 param get /component pose_values` you get a message back that says `Parameter not set.`.

**Comment by domire8** [comment]
> The consequence is that "default empty" parameters cannot be validated automatically the first time they are set, which is acceptable but will need to be made clear in the documentation later. Any validation would have to occur "in place" where you get and use the parameter value.

Quite true, maybe we can resolve that by moving around that `continue

**Comment by buschbapti** [comment]
I am not sure exactly what was the initial issue (maybe I missed it if we discussed about it). But it seems what you wanted is to declare a parameter without giving it a default value, isn't it? Did you try the solution specified here https://github.com/ros2/rclcpp/blob/master/rclcpp/doc/notes_on_statically_typed_parameters.md ? I guess you might have but just in case to be sure.

**Comment by domire8** [comment]
> I am not sure exactly what was the initial issue (maybe I missed it if we discussed about it). But it seems what you wanted is to declare a parameter without giving it a default value, isn't it? Did you try the solution specified here https://github.com/ros2/rclcpp/blob/master/rclcpp/doc/notes_on_statically_typed_parameters.md ? I guess you might have but just in case to be sure. Maybe something like:
> 
> ```
> self.declare_parameter(ros_param.name, rclpy::ParameterType::PARAMETER_DOUBLE_ARRAY, 
>                                       descriptor=ParameterDescriptor(description=description,
>                                                                                            type=Parameter.Type.DOUBLE_ARRAY))
> ```
> 
> I don't know if that would work in python though as I am not sure about the signature of the function there.

Yeahhh this is getting a bit annyoing, I will have to call a developer standup to clarify those issues...Python and C++ really don't behave the same here..

**Comment by domire8** [comment]
So the current progress is as followed:
- In C++, adding an empty parameter from the modulo interface is okay, getting it from the ROS interface throws an exception, and setting a parameter with a  `NOT_SET` parameter on the ROS interface does undeclare that parameter
- In Python, adding an empty parameter from the modulo interface is okay, getting it from the ROS interface does not throw an exception and setting a parameter with a `NOT_SET` parameter on the ROS interface undeclares that parameter

File: source/modulo_components/include/modulo_components/ComponentInterface.hpp
Comment (buschbapti) [inline source/modulo_components/include/modulo_components/ComponentInterface.hpp:405]: Shouldn't `descriptor` be third argument there?

Comment (buschbapti) [inline source/modulo_components/include/modulo_components/ComponentInterface.hpp:405]: I was looking by commit but it seems it changed line. My comment was for here.

Comment (domire8) [reply source/modulo_components/include/modulo_components/ComponentInterface.hpp:405]: Oh that might solve my problem!

Comment (domire8) [reply source/modulo_components/include/modulo_components/ComponentInterface.hpp:405]: It doesn't...you cannot specify the type as second argument and allow dynamic typing at the same time

**Review by buschbapti — DISMISSED** [review]
This looks good now and I believe well tested. Some minor style comments.

File: source/modulo_components/include/modulo_components/ComponentInterface.hpp
Comment (eeberhard) [inline source/modulo_components/include/modulo_components/ComponentInterface.hpp:471]: This is problematic when `parameter` is empty. Because the ros_parameter is of type string when you add_parameter, new_parameter will _not_ be empty. This means that `validate_parameter` will see a different, non-empty, default-valued version of the parameter. For an integer, this might mean that we want to add a "not set" number, but validate parameter will see a value of `0` instead.

**Review by eeberhard — COMMENTED** [review]
As discussed, I think we have to revert to using the NOT_SET parameter type to make sure the parameter validation can be fully consistent.

**Comment by domire8** [comment]
Close in favor of #97

**Pull request closed without merging**
